ang/civiawards/award-creation/directives/review-panels/review-panels.directive.js

```diff
@@ -13,28 +13,213 @@
   });
 
   module.controller('CiviawardReviewPanelsController', function (
-    $scope, ts, dialogService, crmApi, crmStatus, getSelect2Value) {
+    $q, $scope, ts, dialogService, crmApi, crmStatus, getSelect2Value) {
+    var relationshipTypesIndexed = {};
+    var contactsIndexed = {};
+    var groupsIndexed = {};
+
+    $scope.ts = ts;
     $scope.submitInProgress = false;
+    $scope.isLoading = false;
     $scope.submitButtonClickedOnce = false;
     $scope.relationshipTypes = [];
+    $scope.existingReviewPanels = [];
 
     $scope.openCreateReviewPanelPopup = openCreateReviewPanelPopup;
     $scope.addMoreRelations = addMoreRelations;
     $scope.removeRelation = removeRelation;
+    $scope.handleEditReviewPanel = handleEditReviewPanel;
 
     (function init () {
       resetReviewPanelPopup();
-      getRelationshipsTypes()
-        .then(function (relationshipTypeData) {
-          $scope.relationshipTypes = prepareRelationshipsTypes(relationshipTypeData.values);
-        });
+      handleInitialDataLoad();
     }());
 
+    /**
+     * Handles Initial loading of data from API
+     */
+    function handleInitialDataLoad () {
+      $scope.isLoading = true;
+
+      $q.all({
+        groups: getGroups(),
+        relationships: getRelationshipsTypes(),
+        existingReviewPanels: fetchExistingReviewPanels($scope.awardId)
+      }).then(function (fetchedData) {
+        if (fetchedData.existingReviewPanels && fetchedData.existingReviewPanels.length === 0) {
+          return fetchedData;
+        }
+
+        return fetchContactsFromPanelsAndIndexById(fetchedData.existingReviewPanels)
+          .then(function () {
+            return fetchedData;
+          });
```

Comment:
there are two issues with `fetchContactsFromPanelsAndIndexById`:

1 - it does not return anything. It's doing something hidden and we need to peek at the function to understand it.
2 - it's doing 2 things at the same time: it fetches data, and applies some indexing (even if it's minimal)

I believe the best action would be to append the response of `fetchContactsFromPanels` to `fetchedData`:

```js
.then(function (fetchedData) {
  var shouldFetchContactsFromPanels = fetchedData.existingReviewPanels && 
    fetchedData.existingReviewPanels.length === 0;

  if (!shouldFetchContactsFromPanels) {
    return fetchedData;
  }

  return fetchContactsFromPanels(fetchedData.existingReviewPanels)
    .then(function (contactsFromPanels) {
      return _.extend({}, fetchedData, {
        contactsFromPanels: contactsFromPanels
      });
    });
})
.then(function (fetchedData) {
  contactsIndexed = _.indexBy(fetchedData.contactsFromPanels, 'id');
  groupsIndexed = _.indexBy(fetchedData.groups, 'id');
  // ...
});
```

As a note, if you want to make a specialized function you could have this instead:

```js
fetchContactsFromPanels(someData)
  .then(storeContactsIndexedById);
```